modin/core/storage_formats/pandas/query_compiler.py

```diff
@@ -435,10 +425,17 @@ def where_builder_second_pass(df, new_other, **kwargs):
             def where_builder_series(df, cond):
                 return df.where(cond, other, **kwargs)
 
-            new_modin_frame = self._modin_frame.binary_op(
-                where_builder_series, cond._modin_frame, join_type="left"
+            new_modin_frame = self._modin_frame.nary_op(
+                where_builder_series, [cond._modin_frame], join_type="left"
             )
-        return self.__constructor__(new_modin_frame)
+        result = self.__constructor__(new_modin_frame)
+        # The binary operation may reorder rows or columns, but where requires
+        # that we keep the original index and columns
+        if not result.index.equals(self.index):
+            result = result.reindex(axis=0, labels=self.index)
```

Comment:
`reindex` looks too costly. Could you make some perf measurements to cover this case?

Reply:
I passed `join=how` to the binary operation instead. When indexes are different, perf is not much good because we have to do 2 reindexes each for `other` and `cond` before we can even start the `where`. But that's a problem with binary operations in general. Anyway, perf in that case for the example I give is not worse than pandas.

Reply:
I wonder if we can `_copartition` and `.reindex` at the same time in a single operation... would that be much faster?

Reply:
This comment is outdated now that I took out the `reindex`.